### PR TITLE
support dicts in ReturnnDumpHDFJob

### PR DESCRIPTION
**Pull request by michelwi**
the docstring suggested that data should only be of type `str` while an inline comment suggested `dict|Path|str`.
the old implementation definitively worked for strings - although 
```
if isinstance(data, str):
    data = str(data)
```
seemed kinda weird to me - and otherwise assumed Path - again, I am unsure how a path defines a dataset. 

Anyway, now support for dicts is added while - I think - the handling of str and Path is unchanged.

Kudos to Jan Engler for the `resolve_sis_paths_in_dict` function. I think it might be useful in many places and therefore added it under util.

### CODE REVIEW

File: util.py [hunk 1]
```diff
@@ -279,3 +279,10 @@ def check_file_sha256_checksum(filename, reference_checksum):
     :param str filename: a single file to be checked
     """
     assert compute_file_sha256_checksum(filename) == reference_checksum
+
```
Comment (michelwi) [inline util.py:282]:
```suggestion


```
for Black

File: returnn/hdf.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ def __init__(
     ):
         """
 
-        :param str data: a string defining a RETURNN dataset
+        :param dict|Path|str data: a string defining a RETURNN dataset
```
Comment (michelwi) [inline returnn/hdf.py:36]:
```suggestion
        :param dict|Path|str data: a dict, path, or string defining a RETURNN dataset
```
update docstring

**Review by michelwi — COMMENTED** [review]
some inline changes

File: util.py [hunk 3]
```diff
@@ -279,3 +279,11 @@ def check_file_sha256_checksum(filename, reference_checksum):
...
+
+def resolve_sis_paths_in_dict(dictionary):
```
Comment (michelwi) [inline util.py:284]:
```suggestion
def resolve_sis_paths_in_dict(dictionary):
    """
    Iterates through a (nested) dict and replaces all Path instances 
    with the respective file paths. Only to be used in the Worker.
    
    :param dict dictionary: a dict containing Path objects to be replaced
    """
```
@JackTemaki you like the docstring?

Comment (JackTemaki) [reply util.py:284]: > Only to be used in the Worker

Perfect!

File: util.py [hunk 4]
```diff
@@ -279,3 +279,17 @@ def check_file_sha256_checksum(filename, reference_checksum):
...
```
Comment (vieting) [inline util.py:284]: sisyphus has a function `extract_paths` which does a very similar thing ([here](https://github.com/rwth-i6/sisyphus/blob/master/sisyphus/tools.py#L74)). Maybe that could be used here as well? On the other hand, this here is probably easier to read.

Comment (michelwi) [reply util.py:284]: I know about this function. The problem is, that it flattens the data structure in the process (I.e. Returned is a flat list of all path-strings).
I didn't find an elegant solution to write it back in the dict structure.

Comment (vieting) [reply util.py:284]: Alright, then let's keep it as it is.

**Review by vieting — APPROVED** [review]

**Comment by JackTemaki** [comment]
Looks good, but if we are at it, shouldn't the function also cover lists and tuples as well? And maybe also `Variables`?

**Comment by dthulke** [comment]
Doesn't `instanciate_delayed` do all of this already?
What about moving this function to util?

https://github.com/rwth-i6/i6_core/blob/970cd004a83d7220dd6aa288a72bb93a3a7a6373/returnn/config.py#L21-L28

**Comment by michelwi** [comment]
> Doesn't `instanciate_delayed` do all of this already? What about moving this function to util?

Yes, thanks. Exactly what is needed. I am in favor of moving it to util.py as tis might be used by more than just the returnn config.

**Review by JackTemaki — APPROVED** [review]

**Comment by michelwi** [comment]
`instanciate_delayed` is used in i6_experiments by @JackTemaki and @mmz33 here:
https://github.com/rwth-i6/i6_experiments/search?q=instanciate_delayed

do we have to change the import or will it still work as it is imported in `i6_core.returnn.config`?

**Comment by JackTemaki** [comment]
> `instanciate_delayed` is used in i6_experiments by @JackTemaki and @mmz33 here: https://github.com/rwth-i6/i6_experiments/search?q=instanciate_delayed
> 
> do we have to change the import or will it still work as it is imported in `i6_core.returnn.config`?

As long as it is still imported further imports will work. @vieting can you review again so that we can merge?

**Pull request merged**